### PR TITLE
API Split ownership and recursive publishing into RecursivePublishable

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/silverstripe-framework/issues/7359

Framework docs update at https://github.com/silverstripe/silverstripe-framework/pull/7695

### CODE REVIEW

File: src/RecursivePublishable.php [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        $schema = DataObject::getSchema();
+        $ownedHasMany = array_intersect($owns, array_keys($hasMany));
+        foreach ($ownedHasMany as $relationship) {
+            // Check the owned object is actually versioned and staged
+            $joinClass = $schema->hasManyComponent(get_class($owner), $relationship);
+            $joinInstance = DataObject::singleton($joinClass);
+
+            /** @var Versioned $versioned */
+            $versioned = $joinInstance->getExtensionInstance(Versioned::class);
+            if (!$versioned || !$versioned->hasStages()) {
```
Comment (flamerohr) [inline src/RecursivePublishable.php]: is it not enough to check `Extensible::has_extension($joinClass, Versioned::class)`?

seems a lot more verbose but I don't see the gain in this approach

File: tests/php/VersionedOwnershipTest.php [hunk 2]
```diff
@@ -347,12 +374,93 @@ public function testRecursivePublish()
...
+        /** @var VersionedOwnershipTest\OwnedByUnversioned $book1 */
+        $book1 = $this->objFromFixture(VersionedOwnershipTest\OwnedByUnversioned::class, 'book1');
+        /** @var VersionedOwnershipTest\OwnedByUnversioned $book2 */
+        $book2 = $this->objFromFixture(VersionedOwnershipTest\OwnedByUnversioned::class, 'book2');
+        /** @var VersionedOwnershipTest\OwnedByUnversioned $book3 */
+        $book3 = $this->objFromFixture(VersionedOwnershipTest\OwnedByUnversioned::class, 'book3_published');
+
+        // Remove book3 from draft
+        $book3ID = $book3->ID;
+        $book3->delete();
```
Comment (flamerohr) [inline tests/php/VersionedOwnershipTest.php]: add an assert that `book3` is published

File: src/RecursivePublishable.php (hunk 1 above)
Comment (tractorcow) [reply src/RecursivePublishable.php]: I also realised that the object could be versioned, but without staging enabled (hence the hasStages() check).

File: tests/php/VersionedOwnershipTest.php (hunk 2 above)
Comment (tractorcow) [reply tests/php/VersionedOwnershipTest.php]: Done.

File: src/RecursivePublishable.php (hunk 1 above)
Comment (flamerohr) [reply src/RecursivePublishable.php]: I didn't even remember that was a thing :D

**Comment by filiplikavcan** [comment]
I created simple extension which publishes all assets has_one components that are marked as owned by the non-versioned data object: https://github.com/bratiask/own-assets

**Pull request merged**